File: #0310.minimum-height-trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {

        vector<int> tree[n];
        vector<int> edge(n, 0);
        for (auto e:edges) {
            tree[e[1]].push_back(e[0]);
            tree[e[0]].push_back(e[1]);
            edge[e[0]]++;
            edge[e[1]]++;
        }

        int index = n;
        while (index > 2) {
            queue<int> q;
            for (int i = 0; i < n; i++) {
                if (edge[i] == 1) {
                    q.push(i);
                    edge[i] = -1;
                    index--;
                }
            }
            while (!q.empty()) {
                int f = q.front();
                q.pop();
                for (int g : tree[f]) {
                    edge[g]--;
                }
            }
        }

        vector<int> result;
        for (int i = 0;i < n; i++) {
            if (edge[i] == 0 || edge[i] == 1) {
                result.push_back(i);
            }
        }
        return result;
    }
};

File: #0310.minimum-height-trees.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function(n, edges) {
	if (n === 1) {
        return [0];
    }

	const graphs = edges.reduce((edge, [left, right]) => {
		edge[left] = (edge[left] ?? new Set()).add(right);
		edge[right] = (edge[right] ?? new Set()).add(left);
		return edge;
	}, []);

	let results = graphs.reduce((graph, set, index) => {
		set.size === 1 && graph.push(index);
		return graph;
	}, []);

	while (n > 2) {
		const arr = [];
		n -= results.length;
		results.forEach(node => {
			graphs[node].forEach(value => {
				graphs[value].delete(node);
				graphs[value].size === 1 && arr.push(value);
			});
		});
		results = arr;
	}
	return results;
};
